src/base/emu-i386/simx86/codegen-x86.c

```
@@ -2108,34 +2108,24 @@ static unsigned Exec_x86(unsigned *mem_ref, unsigned long *flg,
 		unsigned char *ecpu, void *SeqStart,
 		unsigned short seqflg, unsigned int *seqbase)
 {
-	unsigned ePC;
-	if (seqflg & F_FPOP) {
-		if (TheCPU.fpstate) {
-			loadfpstate(*TheCPU.fpstate);
-			TheCPU.fpstate = NULL;
-		}
-		/* mask exceptions in generated code */
-		unsigned short fpuc;
-		asm ("fstcw	%0" : "=m"(TheCPU.fpuc));
-		fpuc = TheCPU.fpuc | 0x3f;
-		asm ("fldcw	%0" :: "m"(fpuc));
-	}
-	ePC = Exec_x86_asm(mem_ref, flg, ecpu, SeqStart, seqbase);
 	/* was there at least one FP op in the sequence? */
-	if (seqflg & F_FPOP) {
-		int exs;
-		__asm__ __volatile__ ("fstsw	%0" : "=m"(exs));
-		exs &= 0x7f;
-		if (exs) {
-			e_printf("FPU: error status %02x\n",exs);
-			if ((exs & ~TheCPU.fpuc) & 0x3f) {
-				__asm__ __volatile__ ("fnclex\n" ::: "memory");
-				e_printf("FPU exception\n");
-				TheCPU.err = EXCP10_COPR;
+	if ((seqflg & F_FPOP) && TheCPU.fpstate) {
+		/* mask exceptions in generated code if IGNNE# is set*/
+		uint16_t fpuc = TheCPU.fpstate->cwd;
+		TheCPU.ignne = 0;
+		if (fpu_get_ignne()) {
+			uint16_t fpus = TheCPU.fpstate->swd;
+			TheCPU.ignne = (fpus & ~fpuc) & 0x3f;
+			if (TheCPU.ignne) {
+				e_printf("simx86: TheCPU.ignne = %x\n", TheCPU.ignne);
+				TheCPU.fpstate->cwd ^= TheCPU.ignne;
```

Isn't this cwd software-observable?

---

Yes, that's why it's restored to the original value a few lines below. 

Just the actual host FPU needs to mask the exception, so it gets that mask via loadfpstate (it could have gone that in the old code as well, instead of doing some fp instructions after the loadfpstate)